From 9f44539636116f4e7a8cbd08bd54f2c5c69e3757 Mon Sep 17 00:00:00 2001
From: Luwei Zhou <b45643@freescale.com>
Date: Fri, 29 Nov 2013 12:54:49 +0800
Subject: [PATCH 0643/1074] ENGR00290176 mxc: mlb: Fix MLB suspend/resume
 issue

There is risk that mlb register will be access when clock
is closed. This patch fix it.

Signed-off-by: Luwei Zhou <b45643@freescale.com>
[Original patch taken from git://git.freescale.com/imx/linux-2.6-imx.git]
Signed-off-by: Biyao Zhai <biyao.zhai@windriver.com>
---
 drivers/mxc/mlb/mxc_mlb150.c |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/mxc/mlb/mxc_mlb150.c b/drivers/mxc/mlb/mxc_mlb150.c
index 57708f1..fe0239a 100755
--- a/drivers/mxc/mlb/mxc_mlb150.c
+++ b/drivers/mxc/mlb/mxc_mlb150.c
@@ -2702,14 +2702,14 @@ static int mxc_mlb150_suspend(struct platform_device *pdev, pm_message_t state)
 	struct mlb_data *drvdata = platform_get_drvdata(pdev);
 	struct mlb_dev_info *pdevinfo = drvdata->devinfo;
 
-	mlb150_dev_exit();
-
 	if (pdevinfo && atomic_read(&pdevinfo->on)
 		&& (pdevinfo->fps >= CLK_2048FS))
 		clk_disable_unprepare(drvdata->clk_mlb6p);
 
-	if (pdevinfo && atomic_read(&pdevinfo->opencnt))
+	if (pdevinfo && atomic_read(&pdevinfo->opencnt)) {
+		mlb150_dev_exit();
 		clk_disable_unprepare(drvdata->clk_mlb3p);
+	}
 
 	return 0;
 }
@@ -2719,15 +2719,15 @@ static int mxc_mlb150_resume(struct platform_device *pdev)
 	struct mlb_data *drvdata = platform_get_drvdata(pdev);
 	struct mlb_dev_info *pdevinfo = drvdata->devinfo;
 
-	if (pdevinfo && atomic_read(&pdevinfo->opencnt))
+	if (pdevinfo && atomic_read(&pdevinfo->opencnt)) {
 		clk_prepare_enable(drvdata->clk_mlb3p);
+		mlb150_dev_init();
+	}
 
 	if (pdevinfo && atomic_read(&pdevinfo->on) &&
 		(pdevinfo->fps >= CLK_2048FS))
 		clk_prepare_enable(drvdata->clk_mlb6p);
 
-	mlb150_dev_init();
-
 	return 0;
 }
 #else
-- 
1.7.5.4

